java/src/main/native/include/jni_utils.hpp

```diff
@@ -0,0 +1,1067 @@
+
+  T **data() { return reinterpret_cast<T **>(wrapped.data()); }
+
+  const jlongArray get_jlongArray() const { return wrapped.get_jlongArray(); }
+
+  jlongArray get_jlongArray() { return wrapped.get_jlongArray(); }
+
+  /**
+   * @brief if data has been written back into this array, don't commit
+   * it.
+   */
+  void cancel() { wrapped.cancel(); }
+
+  void commit() { wrapped.commit(); }
+};
+
+/**
+ * @brief wrapper around native_jlongArray to hold pointers that are deleted
+ * if not released, like std::unique_ptr.
+ *
+ * By default any changes to the array will be committed back when
+ * released unless cancel is called first.
+ */
+template <typename T, typename D = std::default_delete<T>> class unique_jpointerArray {
+private:
+  std::unique_ptr<native_jpointerArray<T>> wrapped;
+  D del;
+
+public:
+  unique_jpointerArray(unique_jpointerArray const &) = delete;
+  unique_jpointerArray &operator=(unique_jpointerArray const &) = delete;
+
+  unique_jpointerArray(JNIEnv *const env, jlongArray orig)
+      : wrapped(new native_jpointerArray<T>(env, orig)) {}
+
+  unique_jpointerArray(JNIEnv *const env, jlongArray orig, const D &del)
+      : wrapped(new native_jpointerArray<T>(env, orig)), del(del) {}
+
+  unique_jpointerArray(JNIEnv *const env, int len)
+      : wrapped(new native_jpointerArray<T>(env, len)) {}
+
+  unique_jpointerArray(JNIEnv *const env, int len, const D &del)
+      : wrapped(new native_jpointerArray<T>(env, len)), del(del) {}
+
+  unique_jpointerArray(JNIEnv *const env, T *arr, int len)
+      : wrapped(new native_jpointerArray<T>(env, arr, len)) {}
+
+  unique_jpointerArray(JNIEnv *const env, T *arr, int len, const D &del)
+      : wrapped(new native_jpointerArray<T>(env, arr, len)), del(del) {}
+
+  bool is_null() const noexcept { return wrapped == NULL || wrapped->is_null(); }
+
+  int size() const noexcept { return wrapped == NULL ? 0 : wrapped->size(); }
+
+  void reset(int index, T *new_ptr = NULL) {
+    if (wrapped == NULL) {
+      throw std::logic_error("using unique_jpointerArray after release");
+    }
+    T *old = (*wrapped)[index];
+    if (old != new_ptr) {
+      (*wrapped)[index] = new_ptr;
+      del(old);
+    }
+  }
+
+  T *get(int index) {
+    if (wrapped == NULL) {
+      throw std::logic_error("using unique_jpointerArray after release");
+    }
+    return (*wrapped)[index];
+  }
+
+  T *const *get() {
+    if (wrapped == NULL) {
+      throw std::logic_error("using unique_jpointerArray after release");
+    }
+    return wrapped->data();
+  }
+
+  jlongArray release() {
+    if (wrapped == NULL) {
+      return NULL;
+    }
+    wrapped->commit();
+    jlongArray ret = wrapped->get_jlongArray();
+    wrapped.reset(NULL);
+    return ret;
+  }
+
+  ~unique_jpointerArray() {
+    if (wrapped != NULL) {
+      for (int i = 0; i < wrapped->size(); i++) {
+        reset(i, NULL);
+      }
+    }
+  }
+};
+
+/**
+ * @brief RAII for jintArray to be sure it is handled correctly.
```

Comment:
You have multiple RAII wrappers for different types -- similar types like long and int. Could this be replaced by a single template class?

Reply:
The JNI API is not really a C++ API.  It is a C API with some magic that lets you call some functions like C++ methods.  So we would need a way to drop in the proper function for each subclass `GetLongArrayElements`, `GetIntArrayElements`, etc in each instantiation of the template.  I think we could do it, but it is not simple.  Happy to explore it more in a follow-on issue.

Reply:
I made it work.  Not sure it is the cleanest implementation.

f9a73fb and 45eeeed